### PR TITLE
Replicas and shards for mysql and postgres storages

### LINKED ISSUE #20969: Allow replicas in MySQL engine
It is a good idea to allow several replicas in MySQL table engine.

### PR DESCRIPTION
**Pull request by kssenii**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support for replicas in MySQL/PostgreSQL table engine / table function. Added wrapper storage over MySQL / PostgreSQL storages to allow shards. Closes https://github.com/ClickHouse/ClickHouse/issues/20969.

### CODE REVIEW

File: src/Dictionaries/MySQLDictionarySource.cpp [hunk 1]
```diff
@@ -119,67 +119,43 @@ std::string MySQLDictionarySource::getUpdateFieldAndDate()
     }
 }
 
-BlockInputStreamPtr MySQLDictionarySource::retriedCreateMySqlBIStream(const std::string & data_fetch_query_str, const size_t max_tries)
+BlockInputStreamPtr MySQLDictionarySource::loadBase(const String & query)
```
Comment (tavplubix) [inline src/Dictionaries/MySQLDictionarySource.cpp:122]: The name is not clear

File: src/Formats/MySQLBlockInputStream.h [hunk 2]
```diff
@@ -52,22 +53,27 @@ class MySQLBlockInputStream : public IBlockInputStream
 
 /// Like MySQLBlockInputStream, but allocates connection only when reading is starting.
 /// It allows to create a lot of stream objects without occupation of all connection pool.
-class MySQLLazyBlockInputStream final : public MySQLBlockInputStream
+/// Also mekes attempts to reconnect in case of connection failures.
```
Comment (tavplubix) [inline src/Formats/MySQLBlockInputStream.h:56]: Typo

File: src/Storages/PostgreSQL/PostgreSQLPoolWithFailover.h [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+#include <Core/Types.h>
+#include <Poco/Util/AbstractConfiguration.h>
+#include "PostgreSQLConnectionPool.h"
+
+
+namespace postgres
+{
+
+class PoolWithFailover
```
Comment (tavplubix) [inline src/Storages/PostgreSQL/PostgreSQLPoolWithFailover.h:11]: It's very similar to `mysqlxx::PoolWithFailover`, do you consider extracting common code to some generic abstraction?

**Comment by tavplubix** [comment]
`test_mysql_protocol/test.py::test_mysql_client_exception` failure is probably related

File: base/mysqlxx/PoolWithFailover.h [hunk 4]
```diff
@@ -6,11 +6,14 @@
...
 #define MYSQLXX_POOL_WITH_FAILOVER_DEFAULT_MAX_CONNECTIONS 16
 #define MYSQLXX_POOL_WITH_FAILOVER_DEFAULT_MAX_TRIES 3
+#define MYSQLXX_POOL_WITH_FAILOVER_DEFAULT_MAX_ADDRESSES 1
 
 
 namespace mysqlxx
 {
     /** MySQL connection pool with support of failover.
+      *
+      * For dictionary source:
```
[the remark this answers is not in the document]
Comment (kssenii) [reply base/mysqlxx/PoolWithFailover.h:16]: Yes, it is just a part of comments for dictionary. Part of comments for storage is below.

File: src/Dictionaries/MySQLDictionarySource.cpp (hunk 1 above)
Comment (kssenii) [reply src/Dictionaries/MySQLDictionarySource.cpp:122]: I made its name the same as it was made for XDBCDictionarySource, and then in PostgreSQL in another PR. So here it should better also be the same.

**Comment by UnamedRus** [comment]
Does it support case when we have only one dns name with multiple ip's for mysql servers for it?

**Comment by kssenii** [comment]
@UnamedRus, it depends on what behaviour you expect. It should work, but the addresses might be cached.

**Comment by UnamedRus** [comment]
Clickhouse would understand that each IP address is different replica, so it would be able to use all replicas for LA and HA purposes.

**Review by tavplubix — APPROVED** [review]

**Comment by tavplubix** [comment]
Functional stateful tests (release, DatabaseReplicated) - `All connection tries failed while connecting to ZooKeeper. nodes: [::1]:29181, [::1]:9181, [::1]:19181`
Functional stateless tests (release, wide parts enabled) - `Session expired`
Performance - unrelated
Yandex synchronization check - was broken in master
Integration tests flaky check (asan) - `test_mysql_protocol/test.py::test_mysql_federated` - AFAIK it used to be flaky earlier, so we can ignore it and fix in another PR

Integration tests - some failures (such as `test_postgresql_protocol`) look suspicious, we should investigate it before merging

**Comment by gyuton** [comment]
Internal documentation ticket: DOCSUP-8324.

**Comment by shacuros** [comment]
Thanks for the feature, this could simplify some parts in our system.

I tested the functionality with a sharded setup and get an error when trying it as Tablefunction. When using it as Tableengine it works as it should:



```
DESCRIBE TABLE mysql('db-shard-4:3306', 'Datamarts', 'Account_Clicks', 'user***', 'password***');

┌─name───────────────┬─type─────────────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
│ day                │ Date             │              │                    │         │                  │                │
│ uniqueAccountId    │ UInt32           │              │                    │         │                  │                │
│ number             │ UInt32           │              │                    │         │                  │                │
│ cost               │ String           │              │                    │         │                  │                │
└────────────────────┴──────────────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘

```

Selecting from single shards manually:

```
select count(*), sum(number) from mysql('db-shard-4:3306', 'Datamarts', 'Account_Clicks', 'user***', 'password***') where day >= '2021-01-01' and day < '2021-04-01';

┌─count()─┬─sum(number)─┐
│    1275 │       76480 │
└─────────┴─────────────┘

select count(*), sum(number) from mysql('db-shard-10:3306', 'Datamarts', 'Account_Clicks', 'user***', 'password***') where day >= '2021-01-01' and day < '2021-04-01';

┌─count()─┬─sum(number)─┐
│    7866 │     9265808 │
└─────────┴─────────────┘

```

Selecting if it would be replicas:

```
select count(*), sum(number) from mysql('db-shard-{4|10}:3306', 'Datamarts', 'Account_Clicks', 'user***', 'password***') where day >= '2021-01-01' and day < '2021-04-01';
┌─count()─┬─sum(number)─┐
│    7866 │     9265808 │
└─────────┴─────────────┘

```
Selecting as Shards
```
select count(*), sum(number) from mysql('db-shard-{4|10}:3306', 'Datamarts', 'Account_Clicks', 'user***', 'password***') where day >= '2021-01-01' and day < '2021-04-01';

Received exception from server (version 21.5.6):
Code: 1000. DB::Exception: Received from localhost:9000. DB::Exception: Exception: Connections to all replicas failed: Datamarts@db-shard-{4,10}:3306 as user user***


```
Creating an ExternalDistributed table works fine:

```
CREATE TABLE Daily_Account_Clicks
(
    `day` Date,
    `number` UInt32,
    `uniqueAccountId` UInt32
)
ENGINE = ExternalDistributed('MySQL', 'db-shard-{4,10}:3306', 'Datamarts', 'Account_Clicks', 'user***', 'password***');

SELECT
    count(*),
    sum(number)
FROM Daily_Account_Clicks
WHERE (day >= '2021-01-01') AND (day < '2021-04-01')

Query id: e4d7036d-990e-4a1a-861e-ac87d391dcdc

┌─count()─┬─sum(number)─┐
│    9141 │     9342288 │
└─────────┴─────────────┘




```

**Comment by kssenii** [comment]
@shacuros you mean trying to create table function externalDistributed? It just was implemented only as table engine. But table function is not difficult to implement also..

**Comment by shacuros** [comment]
@kssenii According to the changelog it should already work with the mysql table function, as the example was given in the change log:

"Added support for replicas and shards in MySQL/PostgreSQL table engine / table function. You can write SELECT * FROM mysql('host{1,2}-{1|2}', ...). Closes #20969. #22217 (Kseniia Sumarokova)."

But as you can see in my code examples, when using comata for shard separation, it produces an error message, that it cannot connect to all replicas.
Having a workaround to use ExternalDistributed for table function as well would be suitable i think, but then the error message should also be adjusted.

I'd suggest adding ExternalDistributed table function, and when using mysql table function with shards instead of only replicas to use an errormessage like that:

"Sharded connections are not possible for MySQL Tablefunction, use ExternalDistributed instead"

**Comment by kssenii** [comment]
@shacuros  Ok, its just changelog was a little incorrect: replicas were added for both table engine / table function. Shards and replicas combination only as table engine.

But I will add this support for table function also :)

**Pull request merged**